File: src/main/java/com/twibright/optar/Optar.java
package com.twibright.optar;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;

/**
 * Optar encoder. Produces one or more P5 PGM pages per input file.
 *
 * CLI: optar &lt;input file&gt; [filename base]
 * Produces {base}_0001.pgm ... {base}_NNNN.pgm where base defaults to "optar_out"
 * and also becomes the human-readable file label at the bottom of every page.
 */
public final class Optar {

    private static final int WIDTH       = Common.WIDTH;
    private static final int DATA_WIDTH  = Common.DATA_WIDTH;
    private static final int DATA_HEIGHT = Common.DATA_HEIGHT;
    private static final int TEXT_WIDTH  = Common.TEXT_WIDTH;
    private static final int TEXT_HEIGHT = Common.TEXT_HEIGHT;
    private static final int BORDER      = Common.BORDER;
    private static final int CHALF       = Common.CHALF;
    private static final int CPITCH      = Common.CPITCH;
    private static final int HEIGHT      = 2 * BORDER + DATA_HEIGHT + TEXT_HEIGHT;

    private final byte[] ary = new byte[WIDTH * HEIGHT];
    private final int[]  xy  = new int[2];

    private final String base;
    private final String label;
    private final int    nPages;

    private int pageNumber;
    // BCH(63,45) needs a 46-bit accumulator (45 data bits + sentinel) — long.
    private long payloadAccu = 1L;
    private int  symbolIndex;

    private Optar(String base, String label, int nPages) {
        this.base   = base;
        this.label  = label;
        this.nPages = nPages;
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.err.println(
                "Usage: optar <input file> [filename base]\n\n" +
                "Encodes <input file> as a series of P5 PGM pages named\n" +
                "<base>_0001.pgm ... (base defaults to \"optar_out\"). Print the PGMs\n" +
                "on a 600+ DPI laser printer (use pgm2ps to convert to PostScript first)."
            );
            System.exit(1);
        }
        Path input = Path.of(args[0]);
        String base = args.length >= 2 ? args[1] : "optar_out";

        long size = Files.size(input);
        int nPages = (int) (((size * 8L) + Common.NETBITS - 1) / Common.NETBITS);
        if (nPages < 1) nPages = 1;
        if (nPages > 9999) {
            System.err.println("optar: too many pages - 10,000 or more");
            System.exit(1);
        }

        Optar enc = new Optar(base, base, nPages);
        try (InputStream in = new BufferedInputStream(Files.newInputStream(input))) {
            enc.encode(in);
        }
    }

    private void encode(InputStream in) throws IOException {
        newPage();
        int b;
        while ((b = in.read()) >= 0) {
            writeByte(b);
        }
        for (int i = Common.FEC_SMALLBITS - 1; i > 0; i--) {
            writePayloadBit(0);
        }
        flushPage();
    }

    private void writeByte(int b) {
        for (int bit = 7; bit >= 0; bit--) {
            writePayloadBit((b >> bit) & 1);
        }
    }

    private void writePayloadBit(int bit) {
        payloadAccu = (payloadAccu << 1) | (bit & 1L);
        if ((payloadAccu & (1L << Common.FEC_SMALLBITS)) != 0) {
            long data = payloadAccu & ((1L << Common.FEC_SMALLBITS) - 1);
            long code = Bch.encode(data);
            if (symbolIndex >= Common.FEC_SYMS) {
                newPage();
                symbolIndex = 0;
            }
            for (int shift = Common.FEC_LARGEBITS - 1; shift >= 0; shift--) {
                long seq = symbolIndex + (long) (Common.FEC_LARGEBITS - 1 - shift) * Common.FEC_SYMS;
                writeChannelBit((int) ((code >> shift) & 1L), seq);
            }
            payloadAccu = 1L;
            symbolIndex++;
        }
    }

    private void writeChannelBit(int bit, long seq) {
        byte value = (byte) ((bit & 1) != 0 ? 0x00 : 0xff); // 1 = black, 0 = white
        Common.seq2xy(xy, seq);
        int x = xy[0] + BORDER;
        int y = xy[1] + BORDER;
        ary[x + y * WIDTH] = value;
    }

    private void newPage() {
        if (pageNumber > 0) flushPage();
        if (pageNumber >= 9999) {
            throw new IllegalStateException("too many pages");
        }
        pageNumber++;
        formatPage();
    }

    private void flushPage() {
        String out = String.format("%s_%04d.pgm", base, pageNumber);
        try {
            Pgm.write(Path.of(out), ary, WIDTH, HEIGHT);
        } catch (IOException e) {
            throw new RuntimeException("cannot write " + out + ": " + e, e);
        }
    }

    private void formatPage() {
        Arrays.fill(ary, (byte) 0xff);
        drawBorder();
        drawCrosses();
        drawLabel();
    }

    private void drawBorder() {
        // Top border rows.
        for (int y = 0; y < BORDER; y++) {
            Arrays.fill(ary, y * WIDTH, y * WIDTH + WIDTH, (byte) 0);
        }
        // Left/right columns of the data area.
        for (int y = BORDER; y < BORDER + DATA_HEIGHT; y++) {
            int row = y * WIDTH;
            Arrays.fill(ary, row, row + BORDER, (byte) 0);
            Arrays.fill(ary, row + WIDTH - BORDER, row + WIDTH, (byte) 0);
        }
        // The text strip region gets cleared to black then overwritten by label(); the
        // final BORDER rows stay black.
        int textStart = (BORDER + DATA_HEIGHT) * WIDTH;
        Arrays.fill(ary, textStart, textStart + TEXT_HEIGHT * WIDTH, (byte) 0);
        int bottom = (BORDER + DATA_HEIGHT + TEXT_HEIGHT) * WIDTH;
        Arrays.fill(ary, bottom, bottom + BORDER * WIDTH, (byte) 0);
    }

    private void drawCrosses() {
        int maxY = HEIGHT - TEXT_HEIGHT - BORDER - 2 * CHALF;
        int maxX = WIDTH  - BORDER - 2 * CHALF;
        for (int y = BORDER; y <= maxY; y += CPITCH) {
            for (int x = BORDER; x <= maxX; x += CPITCH) {
                drawCross(x, y);
            }
        }
    }

    private void drawCross(int x, int y) {
        // A cross is a 2*CHALF square: upper-left + lower-right are black (0),
        // upper-right + lower-left are white (0xff).
        for (int r = 0; r < CHALF; r++) {
            int row = (y + r) * WIDTH + x;
            Arrays.fill(ary, row,         row + CHALF,        (byte) 0x00);
            Arrays.fill(ary, row + CHALF, row + 2 * CHALF,    (byte) 0xff);
            int row2 = row + CHALF * WIDTH;
            Arrays.fill(ary, row2,         row2 + CHALF,      (byte) 0xff);
            Arrays.fill(ary, row2 + CHALF, row2 + 2 * CHALF,  (byte) 0x00);
        }
    }

    private void drawLabel() {
        String txt = String.format("  0-%d-%d-%d-%d-%d-%d-%d %d/%d %s",
            Common.XCROSSES, Common.YCROSSES, Common.CPITCH, Common.CHALF,
            Common.FEC_ORDER, Common.BORDER, Common.TEXT_HEIGHT,
            pageNumber, nPages, label);

        // Prefix block: the right-most (font_width - TEXT_WIDTH*(127-' ')) pixels of the
        // font strip, copied to destX=0.
        int prefix = Font.WIDTH - TEXT_WIDTH * (127 - ' ');
        int destX = 0;
        drawFontBlock(destX, TEXT_WIDTH * (127 - ' '), prefix);
        destX = prefix;

        // Label text.
        for (int i = 0; i < txt.length(); i++) {
            char ch = txt.charAt(i);
            if (ch >= ' ' && ch <= 127) {
                drawFontBlock(destX, TEXT_WIDTH * (ch - ' '), TEXT_WIDTH);
                destX += TEXT_WIDTH;
            }
        }
    }

    private void drawFontBlock(int destX, int srcX, int width) {
        if (destX + width > DATA_WIDTH) return;
        for (int y = 0; y < TEXT_HEIGHT; y++) {
            int srcRow  = y * Font.WIDTH + srcX;
            int destRow = (BORDER + DATA_HEIGHT + y) * WIDTH + BORDER + destX;
            System.arraycopy(Font.PIXELS, srcRow, ary, destRow, width);
        }
    }
}

File: optar_orig/optar.c
/* (c) GPL 2007 Karel 'Clock' Kulhavy, Twibright Labs */

#include <assert.h> /* assert */
#include <stdint.h> /* uint64_t */
#include <stdio.h>  /* getchar */
#include <stdlib.h> /* exit */
#include <string.h> /* memcpy */

#define width  font_width
#define height font_height
#include "font.h"
#undef width
#undef height

#include "bch.h"
#include "optar.h"
#include "parity.h"
#define HEIGHT      (2 * BORDER + DATA_HEIGHT + TEXT_HEIGHT)
#define TEXT_HEIGHT 24

static unsigned char ary[WIDTH * HEIGHT];
static unsigned char* file_label = (unsigned char*)""; /* The filename written in the
                                         file_label */
static char* output_filename;                          /* The output filename */
static unsigned output_filename_buffer_size;
static unsigned char* base = (unsigned char*)"optar_out"; /* Output filename base */
static unsigned file_number;
FILE* output_stream;
FILE* input_stream;
unsigned n_pages; /* Number of pages calculated from the file length */

void dump_ary(void)
{
    fprintf(output_stream,
        "P5\n"
        "%u %u\n"
        "255\n",
        WIDTH, HEIGHT);

    fwrite(ary, sizeof(ary), 1, output_stream);
}

/* Only the LSB is significant. Writes hamming-encoded bits. The sequence number
 * must not be out of range! */
void write_channelbit(unsigned char bit, unsigned long seq)
{
    int x, y; /* Positions of the pixel */

    bit &= 1;
    bit = -bit;
    bit = ~bit;          /* White=bit 0, black=bit 1 */
    seq2xy(&x, &y, seq); /* Returns without borders! */
    x += BORDER;
    y += BORDER;
    ary[x + y * WIDTH] = bit;
    seq++;
}

void border(void)
{
    unsigned c;
    char* ptr = (char*)(void*)ary;

    memset(ptr, 0, BORDER * WIDTH);
    ptr += BORDER * WIDTH;
    for (c = DATA_HEIGHT; c; c--) {
        memset(ptr, 0, BORDER);
        ptr += WIDTH;
        memset(ptr - BORDER, 0, BORDER);
    }
    memset(ptr, 0, TEXT_HEIGHT * WIDTH);
    ptr += TEXT_HEIGHT * WIDTH;
    /* BORDER bytes into the bottom border */
    memset(ptr, 0, BORDER * WIDTH);
}

void cross(x, y)
{
    unsigned char* ptr = ary + y * WIDTH + x;
    unsigned c;

    for (c = CHALF; c; c--, ptr += WIDTH) {
        memset(ptr, 0, CHALF);
        memset(ptr + CHALF, 0xff, CHALF);
        memset(ptr + CHALF * WIDTH, 0xff, CHALF);
        memset(ptr + CHALF * (WIDTH + 1), 0, CHALF);
    }
}

void crosses(void)
{
    unsigned x, y;

    for (y = BORDER; y <= HEIGHT - TEXT_HEIGHT - BORDER - 2 * CHALF; y += CPITCH)
        for (x = BORDER; x <= WIDTH - BORDER - 2 * CHALF; x += CPITCH)
            cross(x, y);
}

/* x is in the range 0 to DATA_WIDTH-1 */
void text_block(destx, srcx, width)
{
    int x, y;
    unsigned char* srcptr;
    unsigned char* destptr;

    if (destx + width > DATA_WIDTH)
        return; /* Letter doesn't fit */

    srcptr  = (unsigned char*)(void*)header_data + srcx;
    destptr = ary + WIDTH * (BORDER + DATA_HEIGHT) + BORDER + destx;

    for (y = 0; y < TEXT_HEIGHT; y++, srcptr += font_width, destptr += WIDTH) {
        for (x = 0; x < width; x++) {
            destptr[x] = header_data_cmap[srcptr[x]][0] & 0x80 ? 0xff : 0;
        }
    }
}

void label(void)
{
    unsigned x = 0;
    static char txt[DATA_WIDTH / TEXT_WIDTH];
    unsigned char* ptr;
    unsigned txtlen;

    snprintf(txt, sizeof txt, "  0-%u-%u-%u-%u-%u-%u-%u %u/%u %s", XCROSSES, YCROSSES, CPITCH, CHALF, FEC_ORDER, BORDER,
        TEXT_HEIGHT, file_number, n_pages, (char*)(void*)file_label);
    txtlen = strlen((char*)(void*)txt);

    assert(font_height == TEXT_HEIGHT);
    x = font_width - TEXT_WIDTH * (127 - ' ');
    text_block(0, TEXT_WIDTH * (127 - ' '), x);
    for (ptr = (unsigned char*)(void*)txt; ptr < (unsigned char*)(void*)txt + txtlen; ptr++) {
        if (*ptr >= ' ' && *ptr <= 127) {
            text_block(x, TEXT_WIDTH * (*ptr - ' '), TEXT_WIDTH);
            x += TEXT_WIDTH;
        }
    }
}

void format_ary(void)
{
    memset(ary, 0xff, sizeof(ary)); /* White */
    border();
    crosses();
    label();
}

/* Always formats ary. Dumps it if it's not the first one. */
void new_file(void)
{
    if (file_number) {
        dump_ary();
        fclose(output_stream);
    }
    if (file_number >= 9999) {
        fprintf(stderr, "optar: too many pages - 10,000 or more\n");
        exit(1);
    }
    snprintf(output_filename, output_filename_buffer_size, "%s_%04u.pgm", (char*)(void*)base, ++file_number);
    output_stream = fopen(output_filename, "w");
    if (!output_stream) {
        fprintf(stderr, "optar: cannot open %s for writing.\n", output_filename);
        exit(1);
    }
    format_ary();
}

/* Encode the next FEC_SMALLBITS-bit data word with BCH and write its
 * FEC_LARGEBITS channel bits interleaved across the page. */
void write_payloadbit(unsigned char bit)
{
    static uint64_t accu = 1;
    static unsigned long fec_symbol;

    accu = (accu << 1) | (bit & 1);
    if (accu & ((uint64_t)1 << FEC_SMALLBITS)) {
        uint64_t data = accu & (((uint64_t)1 << FEC_SMALLBITS) - 1);
        uint64_t code = bch_encode(data);

        if (fec_symbol >= FEC_SYMS) {
            /* Page full; start a new one. */
            new_file();
            fec_symbol = 0;
        }

        for (int shift = FEC_LARGEBITS - 1; shift >= 0; shift--) {
            write_channelbit(
                (unsigned char)((code >> shift) & 1ULL),
                fec_symbol + (FEC_LARGEBITS - 1 - shift) * FEC_SYMS);
        }
        accu = 1;
        fec_symbol++;
    }
}

void write_byte(unsigned char c)
{
    int bit;

    for (bit = 7; bit >= 0; bit--)
        write_payloadbit(c >> bit);
}

/* Prints the text at the bottom */
/* Makes one output file. */
void feed_data(void)
{
    int c;

    while ((c = fgetc(input_stream)) != EOF) {
        write_byte(c);
    }

    /* Flush the FEC with zeroes */
    for (c = FEC_SMALLBITS - 1; c; c--) {
        write_payloadbit(0);
    }

    dump_ary();
    fclose(output_stream);
}

void open_input_file(char* fname)
{
    input_stream = fopen(fname, "r");
    if (!input_stream) {
        fprintf(stderr, "optar: cannot open input file %s: ", fname);
        perror("");
        exit(1);
    }
    if (fseek(input_stream, 0, SEEK_END)) {
        fprintf(stderr, "optar: cannot seek to the end of %s: ", fname);
        perror("");
        exit(1);
    }
    n_pages = (((unsigned long)ftell(input_stream) << 3) + NETBITS - 1) / NETBITS;
    if (fseek(input_stream, 0, SEEK_SET)) {
        fprintf(stderr, "optar: cannot seek to the beginning of %s: ", fname);
        perror("");
        exit(1);
    }
}

/* argv format:
 * 1st arg - input file
 * 2nd arg(optional) - label and output filename base */
int main(int argc, char** argv)
{

    if (argc < 2) {
        fprintf(stderr, "\n"
                        "Usage: optar <input file> [filename base]\n"
                        "\n"
                        "Will take the input file as data payload and produce optar_out_????.pgm"
                        " which contain the input file encoded onto paper, with error"
                        " correction codes, and automatically split into multiple"
                        " files when necessary. Those pgm's are supposed to be printed"
                        " on laser printer at least 600 DPI for example using GIMP, or use the included pgm2ps to"
                        " convert them to PostScript and print for example"
                        " using a PostScript viewer program.\n"
                        "\n");
        exit(1);
    }
    bch_init();
    open_input_file(argv[1]);

    if (argc >= 3)
        file_label = base = (void*)argv[2];
    output_filename_buffer_size = strlen((char*)(void*)base) + 1 + 4 + 1 + 3 + 1;
    output_filename             = malloc(output_filename_buffer_size);
    if (!output_filename) {
        fprintf(stderr, "Cannot allocate output filename\n");
        exit(1);
    }
    new_file();
    feed_data();
    fclose(input_stream);
    free(output_filename);
    return 0;
}
